This notebook sheds light on the action packed in the following line:
https://github.com/miguelgrinberg/flack/blob/e9baf5ade232d746bad5232f0055093f4dde56b4/flack.py#L89

For convenience, that line is:
```
self.token = binascii.hexlify(os.urandom(32)).decode('utf-8')
```

In [1]:
import binascii
import os

In [2]:
bytes_obj = os.urandom(10)

print(type(bytes_obj))
print(bytes_obj)

<class 'bytes'>
b'M|<?\x7f\x8c0\xb7&D'


In [3]:
# As per https://docs.python.org/3/library/functions.html#func-bytes
# `bytes_obj` is an immutable sequence of integers in the range [0, 256)
print([x for x in bytes_obj])
print(len(bytes_obj))

[77, 124, 60, 63, 127, 140, 48, 183, 38, 68]
10


In [4]:
bytes_obj_in_hex_repr = binascii.hexlify(bytes_obj)

print(type(bytes_obj_in_hex_repr))
print(bytes_obj_in_hex_repr)

<class 'bytes'>
b'4d7c3c3f7f8c30b72644'


Each of cells [2] and [4] has printed one `bytes` object.
- The former represents values in [0, 127] by means of ASCII characters, and values over 127 by means of "the appropriate escape sequence" (e.g. in hexadecimal format such as `b'\xFF'`)
- The latter represents values only by means of "the hexadecimal digits" (i.e. by means of `b'0'` through `b'9'` plus `b'a'` through `b'f'`)

------

Note that a byte in the former always corresponds to 2 bytes in the latter.

------

For example, in accordance with `static/ASCII-Table.svg.png`, we have the following correspondence pairs:
- A byte in "ASCII-character format" in the former (such as `u`) corresponds to two bytes in the latter (such as `'75'`).
- A byte in hexadecimal format in the former (such as `b'\xff'`) corresponds to two bytes in the latter (such as `b'ff'`).
- A byte in "escaped format" in the former (such as `'\n'`) corresponds to two bytes in the latter (such as `'0a'`).

In [5]:
print(len(bytes_obj_in_hex_repr))

20
